alibi_detect/od/pytorch/svm.py

```diff
@@ -0,0 +1,441 @@
+            The proportion of the training data that should be considered outliers. Note that this does
+            not necessarily correspond to the false positive rate on test data, which is still defined when
+            calling the `infer_threshold` method.
+        kernel
+            Kernel function to use for outlier detection.
+        n_components
+            Number of components in the Nystroem approximation, by default uses all of them.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+        """
+        super().__init__(device=device)
+        self.n_components = n_components
+        self.kernel = kernel
+        self.nystroem = _Nystroem(
+            self.kernel,
+            self.n_components
+        )
+        self.nu = nu
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Detect if `x` is an outlier.
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` with leading batch dimension.
+
+        Returns
+        -------
+        `torch.Tensor` of ``bool`` values with leading batch dimension.
+
+        Raises
+        ------
+        ThresholdNotInferredException
+            If called before detector has had `infer_threshold` method called.
+        """
+        scores = self.score(x)
+        if not torch.jit.is_scripting():
+            self.check_threshold_inferred()
+        preds = scores > self.threshold
+        return preds
+
+
+class SgdSVMTorch(SVMTorch):
+    ensemble = False
+
+    def __init__(
+        self,
+        nu: float,
+        kernel: 'torch.nn.Module',
+        n_components: Optional[int] = None,
+        device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
+    ):
+        """SGD Optimization backend for the One class support vector machine (SVM) outlier detector.
+
+        Parameters
+        ----------
+        nu
+            The proportion of the training data that should be considered outliers. Note that this does
+            not necessarily correspond to the false positive rate on test data, which is still defined when
+            calling the `infer_threshold` method.
+        kernel
+            Kernel function to use for outlier detection.
+        n_components
+            Number of components in the Nystroem approximation, by default uses all of them.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+        """
+        if (isinstance(device, str) and device in ('gpu', 'cuda')) or \
+                (isinstance(device, torch.device) and device.type == 'cuda'):
+            warnings.warn(('The `sgd` optimization option is best suited for CPU. If '
```

Comment:
Can we modify to convey that `sgd` will run on cpu regardless and that in this case only the nystroem approximation will be done on gpu.